src/Elasticsearch.Net/Extensions/NameValueCollectionExtensions.cs

```diff
@@ -22,15 +22,17 @@ internal static string ToQueryString(this NameValueCollection nv)
 			var maxLength = 1 + nv.AllKeys.Length - 1; // account for '?', and any required '&' chars
 			foreach (var key in nv.AllKeys)
 			{
-				maxLength += 1 + (key.Length + nv[key]?.Length ?? 0) * 3; // '=' char + worst case assume all key/value chars are escaped
+				var bytes = Encoding.UTF8.GetByteCount(key) + Encoding.UTF8.GetByteCount(nv[key] ?? string.Empty);
+				var maxEncodedSize = bytes * 3; // worst case, assume all bytes are URL escaped to 3 chars
+				maxLength += 1 + maxEncodedSize; // '=' + encoded chars
```

Comment:
I think the `1` character padding could potentially `0` when `nv[key]` is empty. Since we don't write `=`. Not worth the conditional and added complexity though!